### PR TITLE
[CALCITE-6285] Function ARRAY_INSERT produces an incorrect result for…

### PR DESCRIPTION
**Pull request by mihaibudiu**
… negative indices

### CODE REVIEW

**Review by caicancai — APPROVED** [review]
LGTM

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 1]
```diff
@@ -7008,18 +7008,20 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
         "[1, 2, null, 3]", "INTEGER ARRAY NOT NULL");
     f1.checkScalar("array_insert(array[2, 3, 4], 1, 1)",
         "[1, 2, 3, 4]", "INTEGER NOT NULL ARRAY NOT NULL");
-    f1.checkScalar("array_insert(array[1, 3, 4], -2, 2)",
+    f1.checkScalar("array_insert(array[1, 3, 4], -1, 2)",
```
Comment (caicancai) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7011]: I have a question, what will happen if the array length is 5, but I fill in -6? I can test it in spark

Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7011]: @mihaibudiu Hello！
array_insert(array(1, 3, 4), -4, 2) Spark seems to allow this situation, should we add some test samples?
2024-04-02 23-10-05屏幕截图

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 2]
```diff
@@ -7008,18 +7008,20 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
...
+        "[1, 3, 4, 2]", "INTEGER NOT NULL ARRAY NOT NULL");
+    f1.checkScalar("array_insert(array[1, 3, 4], -3, 2)",
         "[1, 2, 3, 4]", "INTEGER NOT NULL ARRAY NOT NULL");
-    f1.checkScalar("array_insert(array[2, 3, null, 4], -5, 1)",
+    f1.checkScalar("array_insert(array[2, 3, null, 4], -6, 1)",
```
Comment (caicancai) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7015]: Maybe we should write array(), spark does not allow array[]

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 1 above)
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7011]: the docs say that this is a prepend. there is a test for this below

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 2 above)
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7015]: these two forms are equivalent

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 1 above)
Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7011]: Thank you for your reply

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 2 above)
Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7015]: thank you for your reply

**Comment by mihaibudiu** [comment]
@chucheng92 I think we agreed to follow the semantics from the newest version of spark, so if you have no objection I will merge this PR.

**Review by chucheng92 — APPROVED** [review]

**Pull request merged**